src/main/java/io/nats/client/MsgDeliveryPool.java

```diff
@@ -0,0 +1,182 @@
+ *  distribution, and is available at http://opensource.org/licenses/MIT
+ */
+
+package io.nats.client;
+
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.locks.*;
+
+
+class MsgDeliveryPool {
+    private List<MsgDeliveryWorker> workers = null;
+    private int                     idx;
+    private boolean                 shutdown;
+
+    // Expand the pool of workers.
+    // This call has no effect for a size lower to 1 or if given size
+    // is lower or equal to current pool size.
+    synchronized void setSize(int size) {
+        if (size <= 0) {
+            return;
+        }
+        if (this.workers == null) {
+            this.workers = new ArrayList<MsgDeliveryWorker>(size);
+        }
+        // We support only expansion of the pool at this point
+        if (size > this.workers.size()) {
+            final int added = size-this.workers.size();
+            for (int i=0; i<added; i++) {
+                MsgDeliveryWorker w = new MsgDeliveryWorker();
+                w.start();
+                this.workers.add(w);
+            }
+        }
+    }
+
+    synchronized int getSize() {
+        if (this.workers == null) {
+            return 0;
+        }
+        return this.workers.size();
+    }
+
+    synchronized void assignDeliveryWorker(AsyncSubscriptionImpl sub) {
+        int idx = this.idx;
+        if (++this.idx >= this.workers.size()) {
+            this.idx = 0;
+            idx = 0;
+        }
+        final MsgDeliveryWorker worker = this.workers.get(idx);
+        sub.setDeliveryWorker(worker);
+    }
+
+    synchronized void shutdown() {
+        if (this.shutdown) {
+            return;
+        }
+        this.shutdown = true;
+        for (int i=0; i<this.workers.size(); i++) {
+            final MsgDeliveryWorker w = this.workers.get(i);
+            w.shutdown();
+        }
+        this.workers.clear();
+    }
+}
+
+class MsgDeliveryWorker extends Thread {
+    private final Lock          mu       = new ReentrantLock();
+    private final List<Message> msgs     = new LinkedList<Message>();
+    private final Condition     cond     = mu.newCondition();
+    private boolean             inWait   = false;
+    private boolean             shutdown = false;
+
+    MsgDeliveryWorker() {
+        this.setName("jnats-msg-delivery-worker-thread");
+    }
+
+    // Add a message to the list and signal the worker thread.
+    // Lock is assumed held on entry.
+    void postMsg(Message msg) {
+        this.msgs.add(msg);
+        if (this.inWait) {
+            this.cond.signal();
+        }
+    }
+
+    @Override
+    public void run() {
+        Message msg = null;
+        AsyncSubscriptionImpl sub = null;
+        ConnectionImpl nc = null;
+        MessageHandler mcb = null;
+        long max = 0;
+        long delivered = 0;
+
+        this.mu.lock();
+        while (true) {
```

Comment:
Is there a reason not to move `!this.shutdown` to the first `while(true)` loop?

Reply:
Because I want to finish dispatching all messages before a shutdown. So shutdown flag itself is not enough condition to exit.